### PR TITLE
Yet More Tool Execution Optimizations

### PR DESCRIPTION
**Pull request by jmchilton**
See individual commits for details.

Builds on the declined #1197 (with fixes from @nsoranzo) with many new optimizations large and small.

### CODE REVIEW

File: config/galaxy.ini.sample [hunk 1]
```diff
@@ -832,6 +832,20 @@ use_interactive = True
...
+# Set the following to a number of threads greater than 1 to spawn
+# a Python task queue for dealing with large tool submissions (either
+# through the tool form or as part of an individual workflow step across
+# large collection). This number of Python threads will be spawned per
+# large request - so you may wish to set it to a number smaller than
+# the core count of your server. This affects workflow scheduling and 
+# web processes, not job handlers.
+#tool_submission_burst_threads = 1
+
+# If the above setting is to set to greater than 1, that number of
```
Comment (martenson) [inline config/galaxy.ini.sample]: You can reword the first sentence as 'If the tool_submission_burst_threads setting above is greater than 0'

File: config/galaxy.ini.sample [hunk 2]
```diff
@@ -832,6 +832,20 @@ use_interactive = True
...
+# a Python task queue for dealing with large tool submissions (either
+# through the tool form or as part of an individual workflow step across
+# large collection). This number of Python threads will be spawned per
+# large request - so you may wish to set it to a number smaller than
+# the core count of your server. This affects workflow scheduling and 
+# web processes, not job handlers.
+#tool_submission_burst_threads = 1
+
+# If the above setting is to set to greater than 1, that number of 
+# threads will be used to process requests resulting in greater than
```
Comment (martenson) [inline config/galaxy.ini.sample]: The rest of the paragraph is cryptic to me.

**Comment by martenson** [comment]
looks great! :+1:

**Comment by jmchilton** [comment]
So the problem was that current datasets do need to be flushed and have an ID before calling the object store create code (sometimes, it actually works in many cases including my manual testing when collecting the timings).

So I have rebased and replaced 619f334 with a1cc98076df21bdca8ee157010ad3dec888ed733 and added a comment about this problem. Unfortunately, this renders the commit message and timings in many subsequent commits incorrect - still I guess they still all improve the situation and shouldn't be broken by an extra flush.

This has also been rebased on top of #1212, so when that is merged the list of commits here will simply shrink to something more manageable.

**Comment by jmchilton** [comment]
Also looks like I need to fix the unit tests.

**Comment by bgruening** [comment]
@jmchilton sorry we also need a rebase here. This looks really awesome!

**Comment by jmchilton** [comment]
Rebased.

**Comment by bgruening** [comment]
:+1: from me as well!

**Pull request merged**